rpcs3/Emu/Cell/SPURecompiler.cpp

```diff
@@ -7162,11 +7190,63 @@ class spu_llvm_recompiler : public spu_recompiler_base, public cpu_translator
 			spu_log.todo("[0x%x] Const SHUFB mask: %s", m_pos, mask);
 		}
 
+		const auto a = get_vr<u8[16]>(op.ra);
+		const auto b = get_vr<u8[16]>(op.rb);
+
+		// Data with swapped endian from a load instruction
+		if (auto [ok, v0] = match_expr(a, byteswap(match<u8[16]>())); ok)
+		{
+			if (auto [ok, v1] = match_expr(b, byteswap(match<u8[16]>())); ok)
+			{
+				// Undo endian swapping, and rely on pshufb to re-reverse endianness
+				const auto x = avg(noncast<u8[16]>(sext<s8[16]>((c & 0xc0) == 0xc0)), noncast<u8[16]>(sext<s8[16]>((c & 0xe0) == 0xc0)));
+				const auto as = byteswap(a);
+				const auto bs = byteswap(b);
+				const auto ax = pshufb(as, c);
+				const auto bx = pshufb(bs, c);
+				set_vr(op.rt4, select(noncast<s8[16]>(c << 3) >= 0, ax, bx) | x);
+				return;
+			}
+
+			if (auto ci = llvm::dyn_cast<llvm::Constant>(b.value))
+			{
+				v128 data = get_const_vector(ci, m_pos, 7000);
+
+				if (data == v128{})
```

Comment:
Supports more variations:
```suggestion
				if (data == std::bit_cast<se_t<v128, true>>(data).load())
```

Reply:
Actually, thinking it over, this isn't quite right. It's only safe when all bytes are equivalent.

For example, if we had a reverseable data of 0xFF000000000000FF, and a mask that starts with 0x0100, we would end up with the incorrect result of 0xFF00 instead of 0x00FF.